### PR TITLE
Darkspear Training Grounds

### PR DESCRIPTION
**Pull request by ghost**
Fix area up a bit allow all classes to complete quest there.

**Changes proposed:**

Darkspear Training Grounds

Replaced Zuni summon spell with newer one.
Scripted newer spell to make zuni spawn in right location.
Added start quest for monk to Jin'thala.
Spawned 3 missing npcs from later expansions, most importantly Monk trainer.
Set all Novices to civilian to prevent combat with Tiki Targets due to sparing not yet emplimented.
Scripted Tiki Targets to have tiki shield, rooted, proper death and respawning.
Scripted The Basics Hitting things so trainer will say text on quest complete.
Fully scripted Proving Pit adding missing classes.
Fixed broken quest chains so now all 9 classes can complete Darkspear Training Grounds quest and move on to Darkspear Hold.

**Issues addressed:**

Closes #  (insert issue tracker number)

**Tests performed:**

Builds and working fine.

**Known issues and TODO list:** (add/remove lines as needed)

- [ The area still needs more work but I don't want to make this PR to big. ]

### CODE REVIEW

**Comment by malcrom** [comment]
Success

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 1]
```diff
@@ -177,9 +178,535 @@ struct npc_mithaka : ScriptedAI
...
+    }
+
+    void SetData(uint32 /*type*/, uint32 data) override
+    {
+        if (data == 1)
+        {
+            me->SetWalk(true);
+
+            if (me->HasStringId("captive_spitescale_scout_one"))
+                me->GetMotionMaster()->MovePoint(0, -1142.49f, -5415.59f, 10.597655f);
```
Comment (mdX7) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:404]: move position to constant variable,
Define a name for 0 in enum (POINT_WHATEVER_DESCRIBES_IT) if you are planning to use it in MovementInform

**Review by mdX7 — CHANGES_REQUESTED** [review]

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 1 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp]: Done

File: sql/updates/world/master/9999_99_99_99_world.sql [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+(31159,512,0,0),
+(31158,512,31159,0),
+(31160,512,31158,0),
+(31161,512,31160,0),
+(31163,512,31161,0);
+
+DELETE FROM `creature_text` WHERE `CreatureID` IN (38037,38245,38246,38244,38242,38247,38243,42618,63310);
+INSERT INTO `creature_text` (`CreatureID`,`GroupID`,`ID`,`Text`,`Type`,`Language`,`Probability`,`Emote`,`Duration`,`Sound`,`BroadcastTextId`,`TextRange`,`comment`) VALUES
+-- Nortet "Warrior Trainer"
+(38037, 0, 0, 'Not bad, $n. Not bad.', 12, 0, 100, 0, 0, 0, 0, 0, 'Nortet'),
```
[the remark this answers is not in the document]
Comment (malcrom) [reply sql/updates/world/master/9999_99_99_99_world.sql]: Done

**Comment by malcrom** [comment]
Changes made.

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 3]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
     bool _shipInPort;
 };
 
+// Echo Isles
+
+class spell_summon_zuni : public SpellScript
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:183]: any class belonging to a given zone_name must have such name in its name, e.g.: spell_durotar_summon_zuni

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 4]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+{
+    void SetDest(SpellDestination& dest)
+    {
+        dest.Relocate({ - 1173.4531f, -5266.401f, 0.85905945f, 0.0f });
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:187]: - must be next to the float

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 5]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+
+struct npc_tiki_target : public ScriptedAI
+{
+    npc_tiki_target(Creature* creature) : ScriptedAI(creature) { _credited = false; }
+
+    void JustAppeared() override
+    {
+        me->SetReactState(REACT_PASSIVE);
+        me->SetTemplateRooted(true);
+        DoCast(me, TiKiTargetMask[urand(0, 2)]);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:215]: DoCastSelf instead of DoCast(me

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 6]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+        DoCast(me, SPELL_ARCANE_MISSILES_TRAINER);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:216]: Same as the above

Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:216]: DoCastSelf instead of DoCast(me

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 7]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+        if (me->GetHealth() <= damage)
+        {
+            damage = 0;
+            me->SetHealth(1);
+
+            if (!_credited)
+            {
+                _credited = true;
+
+                DoCast(me, SPELL_TIKI_TARGET_DEATH);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:230]: DoCastSelf instead of DoCast(me

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 8]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+
+    SAY_SSEND_YOU_TO_YOUR_DEATH   = 0,
+
+    SPELL_FROST_SHOCK             = 15089
+};
+
+Position const PRISIONER_POSITION_ONE = { -1142.49f, -5415.59f, 10.597655f };
+Position const PRISIONER_POSITION_TWO = { -1149.03f, -5526.18f, 8.1045685f };
+
+struct npc_captive_spitescale_scout : public ScriptedAI
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:374]: any class belonging to a given zone_name must have such name in its name

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 9]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+    SPELL_ARCANE_MISSILES_TRAINER = 83470,
+    SPELL_TIKI_TARGET_VISUAL_1    = 71064,
+    SPELL_TIKI_TARGET_VISUAL_2    = 71065,
+    SPELL_TIKI_TARGET_VISUAL_3    = 71066,
+    SPELL_TIKI_TARGET_DEATH       = 71240
+};
+
+uint32 const TiKiTargetMask[3] = { SPELL_TIKI_TARGET_VISUAL_1, SPELL_TIKI_TARGET_VISUAL_2, SPELL_TIKI_TARGET_VISUAL_3 };
+
+struct npc_tiki_target : public ScriptedAI
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:207]: any struct belonging to a given zone_name must have such name in its name

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 10]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+    npc_captive_spitescale_scout(Creature* creature) : ScriptedAI(creature) { }
+
+    void JustAppeared() override
+    {
+        _events.ScheduleEvent(EVENT_UPDATE_JAILOR_GOSSIP, 2s);
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        me->SetUnitFlag(UNIT_FLAG_IMMUNE_TO_PC);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:385]: There's a precious SetImmuneToNPC existing

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 11]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.CancelEvent(EVENT_DESPAWN_OUT_OF_COMBAT);
+        _events.ScheduleEvent(EVENT_CAST_FROSTBOLT, 2s, 4s);
+    }
+
+    void SetData(uint32 /*type*/, uint32 data) override
+    {
+        if (data == 1)
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:402]: magic number

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 12]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+            if (me->HasStringId("captive_spitescale_scout_one"))
+                me->GetMotionMaster()->MovePoint(POINT_PRISIONER_POSITION, PRISIONER_POSITION_ONE);
+            else if (me->HasStringId("captive_spitescale_scout_two"))
+                me->GetMotionMaster()->MovePoint(POINT_PRISIONER_POSITION, PRISIONER_POSITION_TWO);
+        }
+    }
+
+    void MovementInform(uint32 /*type*/, uint32 id) override
+    {
+        if (id == 0)
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:415]: magic number

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 13]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+    {
+        _events.Update(diff);
+
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_UPDATE_JAILOR_GOSSIP:
+                    if (Creature* jailer = me->FindNearestCreature(NPC_DARKSPEAR_JAILOR, 25.0f, true))
+                        jailer->AI()->SetData(1, 1);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:432]: magic numbers

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 14]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+                    break;
+                case EVENT_TALK_TO_PLAYER:
+                    Talk(SAY_SSEND_YOU_TO_YOUR_DEATH);
+                    me->RemoveUnitFlag(UNIT_FLAG_IMMUNE_TO_PC);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:436]: There's a precious SetImmuneToNPC

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 15]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+    { -1158.81f, -5533.08f, 11.939185f }, // Zentabra Pit
+    { -1149.92f, -5407.46f, 13.235063f }, // Voldreka Pit
+    { -1151.54f, -5429.86f, 13.29182f }   // Zabrax Pit
+};
+
+// Facing direction at pit for trainers
+float const EchoIslandTrainersPoints[9] = { 0.2094395f, 3.1764990f, 1.7278759f, 1.4835298f, 4.904375f, 3.3161256f, 0.3141593f, 4.956735f, 1.256637f };
+
+template<uint8 PitPos, uint32 QuestID>
+struct npc_echo_isles_class_trainer : public ScriptedAI
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:511]: any struct belonging to a given zone_name must have such name in its name

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 16]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+        _facingHome = me->GetOrientation();
+    }
+
+    void MovementInform(uint32 /*type*/, uint32 id) override
+    {
+        if (id == 0)
+        {
+            me->SetFacingTo(EchoIslandTrainersPoints[PitPos]);
+        }
+        else if (id == 1)
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:528]: magic number

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 17]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+        else if(me->HasStringId("darkspear_jailor_two"))
+        {
+            _pathCage = PATH_CAGE_TWO;
+            _pathHome = PATH_HOME_TWO;
+        }
+    }
+
+    void SetData(uint32 /*type*/, uint32 data) override
+    {
+        if (data == 1)
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:290]: magic number

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 18]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+            switch (eventId)
+            {
+                case EVENT_MOVE_PIT:
+                    me->SetWalk(true);
+                    me->GetMotionMaster()->MovePoint(0, EchoIslandTrainersPitPoints[PitPos]);
+                    _events.ScheduleEvent(EVENT_MOVE_HOME, 300s);
+                    break;
+                case EVENT_MOVE_HOME:
+                    me->SetWalk(true);
+                    me->GetMotionMaster()->MovePoint(1, _homePosition);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:591]: magic number

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 19]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_PLAYER_ACCEPT_CHALLENGE:
+                    me->GetMotionMaster()->MovePath(_pathCage, false);
+                    break;
+                case EVENT_WALK_BACK_TO_HOME:
+                    if (Creature* scout = me->FindNearestCreature(NPC_CAPTIVE_SPITESCALE_SCOUT, 5.0f, true))
+                        scout->AI()->SetData(1, 1);
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:336]: magic numbers

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 3 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:183]: fixed for next commit.

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 4 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:187]: Fixed for next commit.

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 5 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:215]: Fixed for next commit.

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 6 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:216]: Fixed for next commit.

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 7 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:230]: Fixed for next commit.

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 20]
```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
...
+    PATH_HOME_ONE                 = 3091551,
+    PATH_CAGE_TWO                 = 3090820,
+    PATH_HOME_TWO                 = 3090821,
+
+    SAY_GET_IN_THE_PIT            = 0,
+
+    SPELL_ACTIVATE_DNT            = 227105,
+};
+
+struct npc_darkspear_jailor : public ScriptedAI
```
[the remark this answers is not in the document]
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:267]: Done

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 8 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:374]: Done

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 9 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:207]: Done

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 10 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:385]: It's there once.

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 14 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:436]: done

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 10 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:385]: done

File: src/server/scripts/Kalimdor/zone_durotar.cpp (hunk 15 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_durotar.cpp:511]: done

File: src/server/scripts/Kalimdor/zone_durotar.cpp [hunk 21]
```diff
@@ -264,9 +264,9 @@ enum DarkspearJailor
     SPELL_ACTIVATE_DNT            = 227105,
 };
 
-struct npc_darkspear_jailor : public ScriptedAI
+struct npc_darkspear_jailor_durotar : public ScriptedAI
```
Comment (Nyr97) [inline src/server/scripts/Kalimdor/zone_durotar.cpp:267]: wrong order, it's npc_durotar_darkspear_jailor

**Pull request merged**